File: BackEnd/constants/fast_break_constants.py
"""
Fast Break System Constants

These constants define movement ranges, offsets, and coordinate ranges used throughout
the Fast Break system (DREB → Fast Break, Steal → Fast Break).

All coordinates are in HOME orientation (basket at x=91 for home, x=9 for away).
"""

import random

# Ball Handler Movement (Defensive Stop / Shot Attempt)
BALL_HANDLER_MOVE_X_MIN = 5
BALL_HANDLER_MOVE_X_MAX = 10
BALL_HANDLER_MOVE_Y_RANGE = 3  # ±3 y-coords

# Stopper Positioning (Defensive Stop)
STOPPER_OFFSET_MIN = 1
STOPPER_OFFSET_MAX = 3

# Shot Spot (when ball handler beats defender: shoot from near rim, not confrontation spot)
FB_SHOT_SPOT_X_MIN = 1  # Min x-distance from rim (spots)
FB_SHOT_SPOT_X_MAX = 4  # Max x-distance from rim (spots)
FB_SHOT_SPOT_Y_RANGE = 3  # ±y from rim center

# Defender Positioning (Shot Attempt)
# Primary shot defender sits two x-spots behind the shooter relative to the offense basket.
# Home offense attacks right, so defender x = shooter x - 2.
# Away offense attacks left, so defender x = shooter x + 2.
SHOT_DEFENDER_X_OFFSET = 2
SHOT_DEFENDER_Y_RANGE = 2  # Defender y within ±2 of shooter y
SECONDARY_SHOT_DEFENDER_Y_OFFSET = 3


def fast_break_shot_defender_end_coords(
    shooter_x: float,
    shooter_y: float,
    is_away_offense: bool,
) -> dict:
    """
    Grid destination for the primary fast-break shot contest defender vs the shooter's final spot (HOME grid).

    Rule:
    - Home offense: defender x = shooter x - SHOT_DEFENDER_X_OFFSET
    - Away offense: defender x = shooter x + SHOT_DEFENDER_X_OFFSET
    - Defender y = shooter y + uniform integer in [-SHOT_DEFENDER_Y_RANGE, SHOT_DEFENDER_Y_RANGE]
    """
    x_delta = SHOT_DEFENDER_X_OFFSET if is_away_offense else -SHOT_DEFENDER_X_OFFSET
    defender_x = float(shooter_x) + x_delta
    defender_x = int(max(4, min(97, round(defender_x))))
    defender_y_off = random.randint(-SHOT_DEFENDER_Y_RANGE, SHOT_DEFENDER_Y_RANGE)
    defender_y = int(max(1, min(49, round(float(shooter_y) + defender_y_off))))
    return {"x": defender_x, "y": defender_y}


def fast_break_secondary_shot_defender_end_coords(
    primary_defender_x: float,
    primary_defender_y: float,
    secondary_source_y: float,
) -> dict:
    """
    Grid destination for the secondary fast-break shot defender (beat-stopper path).

    Rule:
    - same x as the primary shot defender
    - y = primary defender y + 3 when the secondary defender starts below that y
      otherwise y = primary defender y - 3
    """
    defender_x = int(max(4, min(97, round(float(primary_defender_x)))))
    y_delta = (
        SECONDARY_SHOT_DEFENDER_Y_OFFSET
        if float(secondary_source_y) > float(primary_defender_y)
        else -SECONDARY_SHOT_DEFENDER_Y_OFFSET
    )
    defender_y = int(max(1, min(49, round(float(primary_defender_y) + y_delta))))
    return {"x": defender_x, "y": defender_y}


# Rebounder Positioning
REBOUNDER_X_MIN = 40
REBOUNDER_X_MAX = 60
REBOUNDER_Y_RANGE = 6  # ±6 y-coords from starting position (defensive stop)

# Shot Attempt Rebounder Positioning
SHOT_ATTEMPT_REBOUNDER_Y_RANGE = 10  # ±10 y-coords from rim (shot attempt)

# Outlet Passer Movement
OUTLET_PASSER_MOVE_X = 7  # Moves forward 7 x-coords toward basket (+7 for home, -7 for away)

# Defensive Stop Determination (legacy y-band — superseded by path cutoff for CR)
DEFENSIVE_STOP_Y_RANGE = 6  # Steal → FB: legacy reference; steal uses after_steal resolver
DEFENSIVE_STOP_Y_RANGE_DREB_OUTLET = 8  # Legacy y-band reference (path cutoff replaces geography gate)

# Unified drive-cutoff tuning (``cutoff_resolution.best_cutoff_on_drive``)
FB_CUTOFF_PATH_CORRIDOR_DREB = 14  # DREB/outlet — wider floor than HCT trap (~11 implicit)
FB_CUTOFF_PATH_CORRIDOR_STEAL = 11  # Legacy steal-entry path through resolve_fast_break_logic
FB_CUTOFF_DEFENDER_TIME_SLACK_DREB = 1.15  # Defenders get 15% arrival-time credit on DREB breaks
FB_CUTOFF_DEFENDER_TIME_SLACK_STEAL = 1.0

# FB Drive Cutoff & Stop Decision (June 2026 — see FB_Drive_Cutoff_Work_Plan.md)
FB_DRIVE_CUTOFF_PATH_CORRIDOR = 14
FB_DRIVE_CUTOFF_TIME_SLACK = 1.0
FB_SHOOT_GEO_RADIUS = 24  # Euclidean to attacking basket OR shoot spot-label exception
FB_POS_O_SHIMMY_MAGNITUDE = 2
FB_CONTEST_MAX_X_TRAIL = 3  # FB-only; pairs with CONTEST_EUCLIDEAN_RADIUS (11)

FB_SHOOT_GEO_SPOT_LABELS = frozenset(
    {"key", "upper midWing", "lower midWing"},
)
FB_PASS_GEO_SPOT_LABELS = frozenset(
    {
        "key",
        "upper midWing",
        "lower midWing",
        "upper wing",
        "lower wing",
        "upper midCorner",
        "lower midCorner",
        "upper corner",
        "lower corner",
    },
)

# Steal Entry Movement (Steal → Fast Break)
STEAL_ENTRY_MOVE_X_MIN = 5  # Minimum x movement toward basket
STEAL_ENTRY_MOVE_X_MAX = 10  # Maximum x movement toward basket
STEAL_ENTRY_MOVE_Y_RANGE = 4  # ±4 y-coords
STEAL_ENTRY_Y_MIN = 3  # Minimum y-coord (clamped)
STEAL_ENTRY_Y_MAX = 47  # Maximum y-coord (clamped)

# Steal HCO Setup Movement (Steal → HCO)
STEAL_HCO_SETUP_MOVE_X_MIN = 3  # Minimum x movement away from basket (ball handler)
STEAL_HCO_SETUP_MOVE_X_MAX = 7  # Maximum x movement away from basket (ball handler)
STEAL_HCO_SETUP_MOVE_Y_RANGE = 3  # ±3 y-coords (ball handler)
STEAL_HCO_SETUP_Y_MIN = 3  # Minimum y-coord (clamped)
STEAL_HCO_SETUP_Y_MAX = 47  # Maximum y-coord (clamped)

# Other Players Movement (Steal HCO Setup - all 9 other players)
STEAL_HCO_SETUP_OTHER_PLAYERS_MOVE_X_MIN = 15  # Minimum x movement toward new offense basket
STEAL_HCO_SETUP_OTHER_PLAYERS_MOVE_X_MAX = 30  # Maximum x movement toward new offense basket
STEAL_HCO_SETUP_OTHER_PLAYERS_MOVE_Y_RANGE = 6  # ±6 y-coords
STEAL_HCO_SETUP_OTHER_PLAYERS_Y_MIN = 4  # Minimum y-coord (clamped)
STEAL_HCO_SETUP_OTHER_PLAYERS_Y_MAX = 46  # Maximum y-coord (clamped)

File: FrontEnd/static/js/phaser/constants/fastBreakConstants.js
/**
 * Fast Break System Constants
 * 
 * These constants define movement ranges, offsets, and coordinate ranges used throughout
 * the Fast Break system (DREB → Fast Break, Steal → Fast Break).
 * 
 * All coordinates are in HOME orientation (basket at x=91 for home, x=9 for away).
 */
import { CLAMP_BOUNDS } from "../animation/courtClamp.js";

/**
 * When false, skips the **"Fast Break!"** entry banners (turn start / FAST_BREAK context).
 * Makes, misses, rebounds, fouls, "Fast Break Score!", "Great Stop!", etc. are unchanged.
 * Runtime override:
 *   window.ENABLE_FAST_BREAK_ENTRY_ANNOUNCEMENTS = false // disable
 *   window.ENABLE_FAST_BREAK_ENTRY_ANNOUNCEMENTS = true  // enable
 */
export const ENABLE_FAST_BREAK_ENTRY_ANNOUNCEMENTS = true;

export function isFastBreakEntryAnnouncementsEnabled() {
  if (typeof window !== "undefined" && typeof window.ENABLE_FAST_BREAK_ENTRY_ANNOUNCEMENTS !== "undefined") {
    return Boolean(window.ENABLE_FAST_BREAK_ENTRY_ANNOUNCEMENTS);
  }
  return ENABLE_FAST_BREAK_ENTRY_ANNOUNCEMENTS;
}

// Ball Handler Movement (Defensive Stop / Shot Attempt)
export const BALL_HANDLER_MOVE_X_MIN = 5;
export const BALL_HANDLER_MOVE_X_MAX = 10;
export const BALL_HANDLER_MOVE_Y_RANGE = 3; // ±3 y-coords

// Stopper Positioning (Defensive Stop)
export const STOPPER_OFFSET_MIN = 1;
export const STOPPER_OFFSET_MAX = 3;

/** FB shot contest: two x-spots behind shooter relative to offense basket, ±Y — matches Python. */
export const SHOT_DEFENDER_X_OFFSET = 2;
export const SHOT_DEFENDER_Y_RANGE = 2;
export const SECONDARY_SHOT_DEFENDER_Y_OFFSET = 3;

/**
 * @param {number} bhX shooter final grid x (HOME)
 * @param {number} bhY shooter final grid y
 * @param {boolean} isHomeOffense
 */
export function fastBreakShotDefenderGridVsShooter(bhX, bhY, isHomeOffense) {
  const xDelta = isHomeOffense ? -SHOT_DEFENDER_X_OFFSET : SHOT_DEFENDER_X_OFFSET;
  let x = bhX + xDelta;
  x = Math.max(CLAMP_BOUNDS.minX, Math.min(CLAMP_BOUNDS.maxX, x));
  const yOff =
    -SHOT_DEFENDER_Y_RANGE +
    Math.floor(Math.random() * (2 * SHOT_DEFENDER_Y_RANGE + 1));
  let y = Math.max(CLAMP_BOUNDS.minY, Math.min(CLAMP_BOUNDS.maxY, bhY + yOff));
  return { x, y };
}

/**
 * @param {number} primaryX primary defender grid x
 * @param {number} primaryY primary defender grid y
 * @param {number} sourceY secondary defender current grid y
 */
export function fastBreakSecondaryShotDefenderGrid(primaryX, primaryY, sourceY) {
  const x = Math.max(CLAMP_BOUNDS.minX, Math.min(CLAMP_BOUNDS.maxX, primaryX));
  const yDelta =
    sourceY > primaryY
      ? SECONDARY_SHOT_DEFENDER_Y_OFFSET
      : -SECONDARY_SHOT_DEFENDER_Y_OFFSET;
  const y = Math.max(CLAMP_BOUNDS.minY, Math.min(CLAMP_BOUNDS.maxY, primaryY + yDelta));
  return { x, y };
}

/** @deprecated Use fastBreakShotDefenderGridVsShooter — kept for any external imports */
export const DEFENDER_X_OFFSET = 6;

// Rebounder Positioning
export const REBOUNDER_X_MIN = 40;
export const REBOUNDER_X_MAX = 60;
export const REBOUNDER_Y_RANGE = 6; // ±6 y-coords from starting position (defensive stop)

// Shot Attempt Rebounder Positioning
export const SHOT_ATTEMPT_REBOUNDER_Y_RANGE = 10; // ±10 y-coords from rim (shot attempt)

/**
 * Minimum grid-x distance from the offense's attacking basket that any
 * non-shooter offense player and any non-shot-defender defense player must
 * keep on Fast Breaks. Applied as a one-sided clamp toward midcourt:
 *   - Home offense (basket at x≈91): target x must be ≤ basket.x − 6 (so ≤ 85).
 *   - Away offense (basket at x≈9):  target x must be ≥ basket.x + 6 (so ≥ 15).
 *
 * Keeps the shot-area visually uncluttered: only the shooter and the primary
 * shot defender(s) (stopper / shot defender) can encroach inside the keep-out
 * zone. Get-back defenders, rebounder-band offense, outlet passer, etc. are
 * all held outside this radius.
 */
export const NON_SHOT_PARTICIPANT_BASKET_KEEPOUT = 6;

// Outlet Passer Movement
export const OUTLET_PASSER_MOVE_X = 7; // Moves forward 7 x-coords toward basket (+7 for home, -7 for away)

// Defensive Stop Determination
export const DEFENSIVE_STOP_Y_RANGE = 6; // Defender must be within ±6 y-coords of outlet receiver to force stop

// Steal Entry Movement (Steal → Fast Break)
export const STEAL_ENTRY_MOVE_X_MIN = 5; // Minimum x movement toward basket
export const STEAL_ENTRY_MOVE_X_MAX = 10; // Maximum x movement toward basket
export const STEAL_ENTRY_MOVE_Y_RANGE = 4; // ±4 y-coords
export const STEAL_ENTRY_Y_MIN = 3; // Minimum y-coord (clamped)
export const STEAL_ENTRY_Y_MAX = 47; // Maximum y-coord (clamped)

// Steal HCO Setup Movement (Steal → HCO)
export const STEAL_HCO_SETUP_MOVE_X_MIN = 3; // Minimum x movement away from basket
export const STEAL_HCO_SETUP_MOVE_X_MAX = 7; // Maximum x movement away from basket
export const STEAL_HCO_SETUP_MOVE_Y_RANGE = 3; // ±3 y-coords
export const STEAL_HCO_SETUP_Y_MIN = 3; // Minimum y-coord (clamped)
export const STEAL_HCO_SETUP_Y_MAX = 47; // Maximum y-coord (clamped)
